Simplify updater logic for downloading and verifying target files

joshuagl opened the pull request:
**Description of the changes being introduced by the pull request**:

The logic for downloading/verifying target files is fairly complex to follow and debug. This PR flattens several internal methods (including a callback) to make the logic simpler to follow.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

joshuagl committed:
updater: remove unused _soft_check_file_length

This internal method isn't used by any code other than tests.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
updater: simplify Updater.download_target() call stack

The call stack and code for download_target() is more complex than
required:
* download_target() : builds target destination filepath, gets length
  and hashes
* _get_target_file() : fixes filenames if consistent snapshots enabled,
  defines verification callback
* _get_file() : iterates mirrors, tries to download files, verifies them

Remove the verification callback and collapse the call stack by a single
level to make the code easier to follow.

Signed-off-by: Joshua Lock <jlock@vmware.com>

sechkova commented on tuf/client/updater.py:
```diff
@@ -1351,24 +1304,44 @@ def _get_target_file(self, target_filepath, file_length, file_hashes,
+      try:
+        file_object = tuf.download.safe_download(file_mirror, file_length)
+
+        # Verify 'file_object' against the expected length and hashes.
```
Do you think it is possible to explain why there is a duplicated check of file length here after the one in `safe_download`?

jku replied:
strictly speaking the checks that are done in functions that safe_download() calls are for the downloaded byte count, whereas _check_file_length() actually looks at the file object. I don't know how that could be of practical difference but...

trishankatdatadog replied:
Good question, but I think better safe than sorry. I think redundant checks are fine as long as no real impact on performance.

jku commented on tuf/client/updater.py:
```diff
@@ -1217,7 +1217,7 @@ def _check_hashes(self, file_object, trusted_hashes):
 
 
-  def _hard_check_file_length(self, file_object, trusted_file_length):
+  def _check_file_length(self, file_object, trusted_file_length):
```
Now that you made me look at this function, I'll comment on it (this may be a bit unrelated to your PR): 

It does 
```
file_object.seek(0)
observed_length = len(file_object.read())
```
which seems wildly unoptimal. I imagine this would be better:
```
file_object.seek(0,2)
observer_length = file_object.tell()
```

JustinCappos replied:
Looks good to me too!  Please add a comment to explain "2" if you add this as it's a magic number here...

joshuagl replied:
Thanks for the suggestion @jku, I pushed an additional patch to implement this optimisation with you listed as co-author.

joshuagl committed:
updater: more optimal file length checking

Rather than read to the end of the file in order to determin its size, use
the whence value of seek() to move the file object's position to the end
of the file, then the tell() method of the file object to read the current
position in bytes.

Co-authored-by: Jussi Kukkonen <jkukkonen@vmware.com>
Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
Trying close/re-open to handle the CI move

trishankatdatadog commented on tuf/client/updater.py:
```diff
@@ -1351,24 +1306,44 @@ def _get_target_file(self, target_filepath, file_length, file_hashes,
+        file_object = None
+
+      else:
+        break
```
Instead of in the `else` here, considering moving the `break` to right after all the checks in `try`.

joshuagl replied:
Nice. Even simpler if we just return `file_object` here and raise the `NoWorkingMirrorError` exception if we exit the loop.

trishankatdatadog commented on tests/test_updater.py:
```diff
@@ -1763,26 +1744,6 @@ def test_11__verify_metadata_file(self):
           metadata_file_object, 'root')
 
 
-  def test_12__get_file(self):
```
Should we replace with a similar test for `_get_target_file()`? Is it tested somewhere else?

joshuagl replied:
`_get_target_file()` is exercised by the existing tests for `updater.download_target()` – not sure it warrants an extra test? `download_target()` only marshals the parameters for `_get_target_file()` and ensures that the directory we fetch into exists.

joshuagl replied:
A test for `_get_target_file()` would only be ~5LOC, so no great effort to add.

trishankatdatadog replied:
As long as the function is covered somewhere, I have no strong objection.

joshuagl committed:
client: simplify loop exit logic

Simplify the loop exit logic in _get_target_file() to simply return a
verified file_object, once we have it, rather than breaking from the loop
and then returning the file_object.

This converts a use of a try/except/else to a try/except and is a little
easier to read.

Signed-off-by: Joshua Lock <jlock@vmware.com>

trishankatdatadog commented:
Why is it called `_get_target_file()`? How are metadata downloaded and verified?

joshuagl commented:
> Why is it called `_get_target_file()`? How are metadata downloaded and verified?

with [`_get_metadata_file()`](https://github.com/theupdateframework/tuf/blob/60dcb95656e8dca156e075dba68ed66715e94c30/tuf/client/updater.py#L1492). It does things slightly differently: doesn't prefix the file with a hash, has different/additional checks for metadata fields (`version` and `spec_version`), checks for expiry, verifies signature, etc.

lukpueh reviewed (approved):
Impeccable work, @joshuagl! 🎉  Thanks for deobfuscating. :)
